shapes/main.cpp

```diff
@@ -1,89 +1,112 @@
-#include <iostream>
-#include <vector>
 #include <algorithm>
-#include <string>
+#include <functional>
+#include <iostream>
+#include <map>
 #include <memory>
-#include "Shape.hpp"
+#include <string>
+#include <type_traits>
+#include <vector>
+#include "Circle.hpp"
 #include "Rectangle.hpp"
+#include "Shape.hpp"
 #include "Square.hpp"
-#include "Circle.hpp"
 
 using namespace std;
 
 using Collection = vector<shared_ptr<Shape>>;
 
-bool sortByArea(shared_ptr<Shape> first, shared_ptr<Shape> second)
-{
-    if(first == nullptr || second == nullptr)
+template <typename T, typename = typename std::enable_if_t<std::is_base_of<Shape, T>::value>>
+void insertIntoCollection(Collection& shapes, shared_ptr<T>& shape) {
+    shapes.push_back(shape);
+}
+
+auto sortByArea = [](shared_ptr<Shape> first, shared_ptr<Shape> second) {
+    if (first == nullptr || second == nullptr) {
         return false;
+    }
     return (first->getArea() < second->getArea());
-}
+};
 
-bool perimeterBiggerThan20(shared_ptr<Shape> s)
-{
-    if(s)
+auto perimeterBiggerThan20 = [](shared_ptr<Shape> s) {
+    if (s) {
         return (s->getPerimeter() > 20);
+    }
     return false;
-}
+};
 
-bool areaLessThan10(shared_ptr<Shape> s)
-{
-    if(s)
-        return (s->getArea() < 10);
+auto areaLessThanX = [x = 10](shared_ptr<Shape> s) {
+    if (s) {
+        return (s->getArea() < x);
+    }
     return false;
-}
+};
 
-void printCollection(const Collection& collection)
-{
-    for (const auto & it : collection)
+void printCollection(const Collection& collection) {
+    for (const auto& it : collection)
         if (it)
             it->print();
 }
 
-void printAreas(const Collection& collection)
-{
-    for (const auto & it : collection)
+void printAreas(const Collection& collection) {
+    for (const auto& it : collection)
         if (it)
             cout << it->getArea() << std::endl;
 }
 
 void findFirstShapeMatchingPredicate(const Collection& collection,
-                                     bool (*predicate)(shared_ptr<Shape> s),
-                                     std::string info)
-{
+                                     std::function<bool(shared_ptr<Shape> s)> predicate,
+                                     std::string info) {
     auto iter = std::find_if(collection.begin(), collection.end(), predicate);
-    if(*iter != nullptr)
-    {
+    if (*iter != nullptr) {
         cout << "First shape matching predicate: " << info << endl;
         (*iter)->print();
-    }
-    else
-    {
+    } else {
         cout << "There is no shape matching predicate " << info << endl;
     }
 }
 
 constexpr int fibo(int n) {
-    if (n<=2) {
+    if (n <= 2) {
         return 1;
     } else {
         return fibo(n - 1) + fibo(n - 2);
     }
 }
 
-int main()
-{
+template <class DerivedType, class... Arguments>
+std::shared_ptr<Shape> make_shape(Arguments&&... args) {
+    return std::shared_ptr<DerivedType>{new DerivedType{std::forward<Arguments>(args)...}};
```

Comment:
```suggestion
    return std::make_shared<DerivedType>(std::forward<Arguments>(args)...);
```
your way is correct, but it's always better to allocate with make_shared